ddtrace/debugging/_probe/model.py

```diff
@@ -79,68 +80,77 @@ def update(self, other: "Probe") -> None:
+class AbstractProbeMixIn(abc.ABC):
+    def __post_init__(self):
+        ...
+
 
-    @limiter.default
-    def _(self):
-        return RateLimiter(
+@dataclass
+class RateLimitMixin(AbstractProbeMixIn):
+    rate: float
+    limiter: RateLimiter = field(init=False, repr=False, compare=False)
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.limiter = RateLimiter(
             limit_rate=self.rate,
             tau=1.0 / self.rate if self.rate else 1.0,
-            on_exceed=lambda: log.warning("Rate limit exceeeded for %r", self),
+            on_exceed=lambda: log.warning("Rate limit exceeded for %r", self),
             call_once=True,
             raise_on_exceed=False,
         )
 
 
-@attr.s
-class ProbeConditionMixin(object):
+@dataclass
+class ProbeConditionMixin(AbstractProbeMixIn):
     """Conditional probe.
 
     If the condition is ``None``, then this is equivalent to a non-conditional
     probe.
     """
 
-    condition = attr.ib(type=Optional[DDExpression])
-    condition_error_rate = attr.ib(type=float, eq=False)
-    condition_error_limiter = attr.ib(type=RateLimiter, init=False, repr=False, eq=False)
+    condition: Optional[DDExpression]
+    condition_error_rate: float = field(compare=False)
+    condition_error_limiter: RateLimiter = field(init=False, repr=False, compare=False)
 
-    @condition_error_limiter.default
-    def _(self):
-        return RateLimiter(
+    def __post_init__(self):
+        super().__post_init__()
+        self.condition_error_limiter = RateLimiter(
             limit_rate=self.condition_error_rate,
             tau=1.0 / self.condition_error_rate if self.condition_error_rate else 1.0,
-            on_exceed=lambda: log.debug("Condition error rate limit exceeeded for %r", self),
+            on_exceed=lambda: log.debug("Condition error rate limit exceeded for %r", self),
             call_once=True,
             raise_on_exceed=False,
         )
 
 
-@attr.s
-class ProbeLocationMixin(object):
+@dataclass
+class ProbeLocationMixin(AbstractProbeMixIn):
     def location(self) -> Tuple[Optional[str], Optional[Union[str, int]]]:
-        """return a tuple of (location,sublocation) for the probe.
-        For example, line probe returns the (file,line) and method probe return (module,method)
+        """Return a tuple of (location, sublocation) for the probe.
+        For example, line probe returns the (file, line) and method probe return (module, method)
         """
         return (None, None)
 
 
-@attr.s
+@dataclass
 class LineLocationMixin(ProbeLocationMixin):
-    source_file = attr.ib(type=Path, converter=_resolve_source_file, eq=False)  # type: ignore[misc]
-    line = attr.ib(type=int, eq=False)
+    source_file: str = field(compare=False)
+    line: int = field(compare=False)
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.resolved_source_file = _resolve_source_file(self.source_file)
```

Comment:
can we add this logic into source_file init ? 
maybe we also like to keep the original value (aka, when we report errors or when _resolve_source_file fails)

Reply:
unfortunately `dataclasses` don't have the ability of using a custom converter so we have to use an extra field here